tests/benchmark/bm_batch_iterator.cpp

```diff
@@ -8,83 +8,98 @@
-                               .hnswParams = HNSWParams{.type = VecSimType_FLOAT32,
-                                                        .dim = dim,
-                                                        .metric = VecSimMetric_Cosine,
-                                                        .initialCapacity = n_vectors + 1,
-                                                        .M = M,
-                                                        .efConstruction = ef_c}};
-        hnsw_index = VecSimIndex_New(&params);
-
-        // Load pre-generated HNSW index.
-        auto location = std::string(getenv("ROOT"));
-        auto file_name = std::string(location) +
-                         "/tests/benchmark/data/DBpedia-n1M-cosine-d768-M64-EFC512.hnsw_v1";
-        auto serializer =
-            hnswlib::HNSWIndexSerializer(reinterpret_cast<HNSWIndex *>(hnsw_index)->getHNSWIndex());
-        serializer.loadIndex(file_name,
-                             reinterpret_cast<HNSWIndex *>(hnsw_index)->getSpace().get());
-        size_t ef_r = 500;
-        reinterpret_cast<HNSWIndex *>(hnsw_index)->setEf(ef_r);
-
-        VecSimParams bf_params = {.algo = VecSimAlgo_BF,
-                                  .bfParams = BFParams{.type = VecSimType_FLOAT32,
-                                                       .dim = dim,
-                                                       .metric = VecSimMetric_Cosine,
-                                                       .initialCapacity = n_vectors}};
-        bf_index = VecSimIndex_New(&bf_params);
-
-        // Add the same vectors to Flat index.
-        for (size_t i = 0; i < n_vectors; ++i) {
-            char *blob =
-                reinterpret_cast<HNSWIndex *>(hnsw_index)->getHNSWIndex()->getDataByInternalId(i);
-            VecSimIndex_AddVector(bf_index, blob, i);
+    ~BM_BatchIterator() {
+        ref_count--;
+        if (ref_count == 0) {
+            VecSimIndex_Free(hnsw_index);
+            VecSimIndex_Free(bf_index);
+            delete queries;
         }
     }
-
-    void TearDown(const ::benchmark::State &state) {
-        VecSimIndex_Free(bf_index);
-        VecSimIndex_Free(hnsw_index);
-    }
-
-    ~BM_BatchIterator() {}
 };
 
+size_t BM_BatchIterator::ref_count = 0;
+BM_BatchIterator *BM_BatchIterator::instance = nullptr;
+
 BENCHMARK_DEFINE_F(BM_BatchIterator, BatchedSearch_BF)(benchmark::State &st) {
 
     size_t batch_size = st.range(0);
     size_t total_res_count = st.range(1);
+    size_t iter = 0;
     for (auto _ : st) {
-        st.PauseTiming();
-        // Generate random query vector before test.
-        std::uniform_real_distribution<> distrib;
-        for (size_t i = 0; i < dim; ++i) {
-            query[i] = (float)distrib(rng);
-        }
-        st.ResumeTiming();
         VecSimBatchIterator *batchIterator =
-            VecSimBatchIterator_New(bf_index, query.data(), nullptr);
+            VecSimBatchIterator_New(bf_index, (*queries)[iter++].data(), nullptr);
```

Comment:
same goes here regarding generality, maybe add a module for the iteration to not overshoot the queries array.
this comment is regarding all queries array access in this file.